### ISSUE TITLE
Smearing for molecules can make SCF unstable. 

### ISSUE THREAD

**Issue by bbusemeyer** [pyscf/pyscf#1195]
I have been trying to use smearing on a Vanadium-containing compound for an numerical experiment, but I found that, although I could converge the molecule without smearing just fine, even a small amount of smearing made the SCF non-convergent. 

I was able to approximately replicate the issue in some simpler atomic calculations. It seems that the issue is somewhat system dependent, from my experimentation. 

(edited to fix a superficial bug)
```
from pyscf.gto import M
from pyscf.scf import ROHF
from pyscf.scf.addons import smearing_
HA = 27.211386024367243

mol = M(
    atom = 'V 0 0 0',  # in Angstrom
    basis = 'ccpvdz',
    spin = 3,
    symmetry = False,
)

myhf = ROHF(mol)
myhf.verbose = 4 
myhf.kernel()

myhf.analyze()

smearhf = smearing_(myhf, sigma=0.001/HA, method='gauss')
smearhf.max_cycle = 200 
smearhf.kernel()

smearhf.analyze()
```

The result of this script, for me, is that the calculation without smearing converges much faster than the one with some 0.1 eV smearing. Even with 0.001 eV smearing, the calculation remains fairly unstable. 

My question is: is this a bug in molecular smearing, or is this an issue with using smearing on this system? Quantum Espresso seems to actually benefit from using smearing on this system, for what it's worth.

**Comment by zhcui** [comment]
I think they may converge to a different solution. Can you check the density matrix?

**Comment by Majokux** [comment]
Hi Brian,

I'm not sure this decorator is well suited for ROHF. The ROHF `make_rdm1` routines uses boolean arrays (i.e. 0 or 1) to build the DM (see  [here](https://github.com/pyscf/pyscf/blob/14b880cdc4bf7bb64f3989a147c0b04622efc498/pyscf/scf/rohf.py#L228)) , and I don't see that this is fixed anyway in the decorator, such that real values for `mo_occ`  are used.

**Comment by bbusemeyer** [comment]
> I think they may converge to a different solution. Can you check the density matrix?

I just had a quick look at the Mulliken pops and the smearing makes the orbitals a bit more diffuse, and it mixes up the d orbitals, but the total number of electrons in the d orbitals is the same. The non-smeared calculation gives integer occupations for the d orbitals. 

> I'm not sure this decorator is well suited for ROHF. 

Should UHF work better? I can try that, I think I did notice in the test cases that the problem seemed to disappear for that. Should ROHF in the PBCs be expected to work?

**Comment by Majokux** [comment]
>Should UHF work better? I can try that, I think I did notice in the test cases that the problem seemed to disappear for that. Should ROHF in the PBCs be expected to work?

For UHF it seems that this decorator mixes between the spin channels (with a common chemical potential), thus destroying the polarization. This is not particular useful I think, instead one should smear alpha- and beta- spin independently, using two different chemical potentials and preserving `cell.spin` as the difference between `sum(mo_occ[0])` and `sum(mo_occ[1])`.

I'm not aware of any problems with ROHF for solids, beyond those which are presents for molecules (non-uniqueness).
However the code in `pbc.scf.krohf` still only seems to allow integer MO occupations.

**Comment by bbusemeyer** [comment]
I see, yeah, actually a bit more testing seems to suggest that a simple switch to UHF didn't actually change much by way of stability. 

Strangely, I just noticed that the problem is exacerbated by a very small smearing ~1e-8 eV, it doesn't converge in 200 iterations.

**Comment by bbusemeyer** [comment]
I haven't seen any problem in solids, so I thought there must be a problem with using it without modifications in molecules, but the presence of `smearing_` in the `pyscf/scf/addons.py` suggested to me that it is supported, hence I raised an issue.

**Comment by zhcui** [comment]
> I haven't seen any problem in solids, so I thought there must be a problem with using it without modifications in molecules, but the presence of `smearing_` in the `pyscf/scf/addons.py` suggested to me that it is supported, hence I raised an issue.

I think it is supported for molecules. But ROHF may not.

**Comment by bbusemeyer** [comment]
> I think it is supported for molecules. But ROHF may not.

Does this mean RHF only? or do you mean something else?

**Comment by zhcui** [comment]
It supports RHF, UHF and GHF

**Comment by bbusemeyer** [comment]
I see, well putting aside the issues that @Majokux raised about the spin not being fixed, I do still see UHF is less stable with smearing applied as well, although its hard to say for certain if the problem is more or less severe than ROHF.

**Comment by bbusemeyer** [comment]
It seems that converting this to a PBC calculation fixes the issue. In the following, the smeared converges in a *fewer* number of iterations (whereas the OBC version has the smeared calculation converging in > 100 iterations). 
```
from pyscf.pbc.gto import Cell
from numpy import eye 
from pyscf.pbc.scf import ROHF
from pyscf.scf.addons import smearing_
HA = 27.211386024367243

cell = Cell()
cell.atom = 'V 0 0 0'
cell.basis = 'ccpvdz'
cell.spin = 3 
cell.a = 12*eye(3)
cell.symmetry = False
cell.build()

myhf = ROHF(cell).density_fit()
myhf.verbose = 4 
myhf.kernel()

smearhf = smearing_(myhf, sigma=1e-8/HA, method='gauss')
smearhf.max_cycle = 200 
smearhf.kernel()
```

**Comment by bbusemeyer** [comment]
The funny thing is that the `make_rdm1` for this PBC ROHF just inherits the `make_rdm1` from the molecular case, I think.

**Comment by sunqm** [comment]
The ROHF calculation converges to a spherical symmetry broken state. Using a relative large sigma, such as 0.01 Hartree, can converge to a result with spherical symmetry. This issue is related to #2289 .  It should have been fixed in PR #2356 .
